File: backend/app/schemas/query_optimization.py
"""
Query Optimization Schemas

Pydantic schemas for query plans, cost estimation, query suggestions,
and performance monitoring.
"""

from typing import Any, Optional
from pydantic import BaseModel, Field
from enum import Enum
from datetime import datetime


# Enums

class QueryPlanNodeType(str, Enum):
    """Types of query plan nodes"""
    SEQ_SCAN = "seq_scan"
    INDEX_SCAN = "index_scan"
    INDEX_ONLY_SCAN = "index_only_scan"
    BITMAP_SCAN = "bitmap_scan"
    NESTED_LOOP = "nested_loop"
    HASH_JOIN = "hash_join"
    MERGE_JOIN = "merge_join"
    SORT = "sort"
    AGGREGATE = "aggregate"
    GROUP_BY = "group_by"
    LIMIT = "limit"
    SUBQUERY = "subquery"
    CTE = "cte"
    UNION = "union"
    MATERIALIZE = "materialize"
    FILTER = "filter"


class QueryComplexity(str, Enum):
    """Query complexity levels"""
    SIMPLE = "simple"
    MODERATE = "moderate"
    COMPLEX = "complex"
    VERY_COMPLEX = "very_complex"


class OptimizationStatus(str, Enum):
    """Optimization status"""
    NOT_ANALYZED = "not_analyzed"
    ANALYZING = "analyzing"
    OPTIMIZED = "optimized"
    NEEDS_ATTENTION = "needs_attention"
    CRITICAL = "critical"


class SuggestionPriority(str, Enum):
    """Suggestion priority levels"""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class SuggestionCategory(str, Enum):
    """Categories of optimization suggestions"""
    INDEX = "index"
    JOIN = "join"
    FILTER = "filter"
    AGGREGATION = "aggregation"
    SUBQUERY = "subquery"
    LIMIT = "limit"
    CACHING = "caching"
    SCHEMA = "schema"
    STATISTICS = "statistics"


class QuerySourceType(str, Enum):
    """Source of the query"""
    CHART = "chart"
    DASHBOARD = "dashboard"
    SQL_LAB = "sql_lab"
    API = "api"
    TRANSFORM = "transform"
    SCHEDULED = "scheduled"
    KODEE = "kodee"


# Query Plan Models

class QueryPlanNode(BaseModel):
    """Node in a query execution plan"""
    id: str
    type: QueryPlanNodeType
    label: str
    table_name: Optional[str] = None
    index_name: Optional[str] = None
    condition: Optional[str] = None
    estimated_rows: int = 0
    actual_rows: Optional[int] = None
    estimated_cost: float = 0.0
    actual_cost: Optional[float] = None
    width: int = 0  # Average row width in bytes
    loops: int = 1
    startup_cost: float = 0.0
    total_cost: float = 0.0
    execution_time_ms: Optional[float] = None
    children: list["QueryPlanNode"] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    is_bottleneck: bool = False


class QueryPlan(BaseModel):
    """Complete query execution plan"""
    id: str
    query_hash: str
    sql: str
    connection_id: str
    root_node: QueryPlanNode
    total_cost: float
    estimated_rows: int
    actual_rows: Optional[int] = None
    execution_time_ms: Optional[float] = None
    planning_time_ms: Optional[float] = None
    complexity: QueryComplexity
    analyzed_at: datetime = Field(default_factory=datetime.utcnow)
    metadata: dict[str, Any] = Field(default_factory=dict)


class QueryPlanRequest(BaseModel):
    """Request to analyze a query plan"""
    sql: str
    connection_id: str
    explain_analyze: bool = False
    explain_buffers: bool = False
    explain_verbose: bool = False


class QueryPlanComparison(BaseModel):
    """Comparison between two query plans"""
    original_plan: QueryPlan
    optimized_plan: QueryPlan
    cost_reduction_percent: float
    row_estimate_diff: float
    execution_time_improvement_ms: Optional[float] = None
    improvements: list[str]


# Cost Estimation Models

class CostComponent(BaseModel):
    """Component of query cost"""
    name: str
    description: str
    estimated_cost: float
    percentage: float
    io_cost: float = 0.0
    cpu_cost: float = 0.0


class CostEstimate(BaseModel):
    """Query cost estimate"""
    query_hash: str
    total_cost: float
    io_cost: float
    cpu_cost: float
    startup_cost: float
    row_estimate: int
    width_estimate: int  # Average row width in bytes
    data_transfer_mb: float
    components: list[CostComponent]
    complexity: QueryComplexity
    estimated_duration_ms: float
    confidence: float  # 0-1, based on statistics freshness


class CostEstimateRequest(BaseModel):
    """Request for cost estimation"""
    sql: str
    connection_id: str
    include_components: bool = True


class ResourceUsageEstimate(BaseModel):
    """Estimated resource usage for a query"""
    memory_mb: float
    temp_space_mb: float
    io_operations: int
    cpu_time_ms: float
    network_transfer_mb: float


# Query Suggestion Models

class QuerySuggestion(BaseModel):
    """Optimization suggestion for a query"""
    id: str
    category: SuggestionCategory
    priority: SuggestionPriority
    title: str
    description: str
    impact: str
    original_snippet: Optional[str] = None
    suggested_snippet: Optional[str] = None
    estimated_improvement_percent: float
    implementation_effort: str  # low, medium, high
    auto_applicable: bool = False


class IndexSuggestion(BaseModel):
    """Suggested index creation"""
    table_name: str
    schema_name: str
    columns: list[str]
    index_type: str = "btree"  # btree, hash, gin, gist
    unique: bool = False
    partial_condition: Optional[str] = None
    estimated_size_mb: float
    estimated_improvement_percent: float
    create_statement: str
    priority: SuggestionPriority


class QueryRewrite(BaseModel):
    """Suggested query rewrite"""
    original_sql: str
    rewritten_sql: str
    rewrite_type: str  # join_order, subquery_to_join, add_index_hint, etc.
    description: str
    estimated_improvement_percent: float
    confidence: float


class OptimizationResult(BaseModel):
    """Result of query optimization analysis"""
    query_hash: str
    original_sql: str
    status: OptimizationStatus
    complexity: QueryComplexity
    cost_estimate: CostEstimate
    suggestions: list[QuerySuggestion]
    index_suggestions: list[IndexSuggestion]
    rewrites: list[QueryRewrite]
    bottlenecks: list[str]
    analyzed_at: datetime = Field(default_factory=datetime.utcnow)


# Query Performance Models

class QueryExecution(BaseModel):
    """Record of query execution"""
    id: str
    query_hash: str
    sql: str
    connection_id: str
    source_type: QuerySourceType
    source_id: Optional[str] = None
    user_id: Optional[str] = None
    workspace_id: Optional[str] = None
    execution_time_ms: float
    rows_returned: int
    rows_scanned: int
    bytes_processed: int
    cached: bool = False
    cache_hit: bool = False
    status: str  # success, error, timeout, cancelled
    error_message: Optional[str] = None
    executed_at: datetime = Field(default_factory=datetime.utcnow)
    metadata: dict[str, Any] = Field(default_factory=dict)


class QueryPerformanceStats(BaseModel):
    """Performance statistics for a query"""
    query_hash: str
    sql_preview: str
    execution_count: int
    total_execution_time_ms: float
    avg_execution_time_ms: float
    min_execution_time_ms: float
    max_execution_time_ms: float
    p50_execution_time_ms: float
    p95_execution_time_ms: float
    p99_execution_time_ms: float
    avg_rows_returned: float
    total_rows_scanned: int
    cache_hit_rate: float
    error_rate: float
    last_executed_at: datetime
    first_executed_at: datetime


class SlowQuery(BaseModel):
    """Slow query record"""
    id: str
    query_hash: str
    sql: str
    connection_id: str
    execution_time_ms: float
    threshold_ms: float
    rows_scanned: int
    rows_returned: int
    source_type: QuerySourceType
    source_id: Optional[str] = None
    user_id: Optional[str] = None
    optimization_status: OptimizationStatus
    suggestions_count: int
    executed_at: datetime


# Query History Models

class QueryHistoryEntry(BaseModel):
    """Entry in query history"""
    id: str
    query_hash: str
    sql: str
    connection_id: str
    source_type: QuerySourceType
    source_id: Optional[str] = None
    user_id: Optional[str] = None
    execution_time_ms: float
    rows_returned: int
    status: str
    executed_at: datetime
    tags: list[str] = Field(default_factory=list)


class QueryHistoryStats(BaseModel):
    """Statistics from query history"""
    total_queries: int
    unique_queries: int
    total_execution_time_ms: float
    avg_execution_time_ms: float
    slow_query_count: int
    error_count: int
    cache_hit_rate: float
    by_source: dict[str, int]
    by_hour: list[dict[str, Any]]
    top_slow_queries: list[SlowQuery]


# Configuration Models

class OptimizationConfig(BaseModel):
    """Query optimization configuration"""
    slow_query_threshold_ms: int = 1000
    auto_analyze: bool = True
    auto_suggest_indexes: bool = True
    max_suggestions_per_query: int = 10
    cache_plans: bool = True
    plan_cache_ttl_seconds: int = 3600
    collect_statistics: bool = True
    statistics_sample_rate: float = 0.1


class OptimizationConfigUpdate(BaseModel):
    """Update optimization configuration"""
    slow_query_threshold_ms: Optional[int] = None
    auto_analyze: Optional[bool] = None
    auto_suggest_indexes: Optional[bool] = None
    max_suggestions_per_query: Optional[int] = None
    cache_plans: Optional[bool] = None
    plan_cache_ttl_seconds: Optional[int] = None
    collect_statistics: Optional[bool] = None
    statistics_sample_rate: Optional[float] = None


# Response Models

class QueryPlanListResponse(BaseModel):
    """List query plans response"""
    plans: list[QueryPlan]
    total: int


class QuerySuggestionListResponse(BaseModel):
    """List query suggestions response"""
    suggestions: list[QuerySuggestion]
    total: int


class SlowQueryListResponse(BaseModel):
    """List slow queries response"""
    queries: list[SlowQuery]
    total: int


class QueryHistoryListResponse(BaseModel):
    """List query history response"""
    entries: list[QueryHistoryEntry]
    total: int


class IndexSuggestionListResponse(BaseModel):
    """List index suggestions response"""
    suggestions: list[IndexSuggestion]
    total: int


# Constants

COMPLEXITY_THRESHOLDS = {
    "simple": {"cost": 100, "joins": 0, "subqueries": 0},
    "moderate": {"cost": 1000, "joins": 2, "subqueries": 1},
    "complex": {"cost": 10000, "joins": 5, "subqueries": 3},
    "very_complex": {"cost": float("inf"), "joins": float("inf"), "subqueries": float("inf")},
}

SLOW_QUERY_THRESHOLDS = {
    "warning": 500,  # ms
    "slow": 1000,  # ms
    "very_slow": 5000,  # ms
    "critical": 30000,  # ms
}

SUGGESTION_TEMPLATES = {
    SuggestionCategory.INDEX: "Consider creating an index on {table}.{columns}",
    SuggestionCategory.JOIN: "Optimize join order: {suggestion}",
    SuggestionCategory.FILTER: "Add filter on {column} to reduce scanned rows",
    SuggestionCategory.AGGREGATION: "Pre-aggregate data or use materialized view",
    SuggestionCategory.SUBQUERY: "Convert subquery to JOIN for better performance",
    SuggestionCategory.LIMIT: "Add LIMIT clause to reduce result set",
    SuggestionCategory.CACHING: "Enable query caching for this frequently-run query",
}


# Helper Functions

def calculate_query_complexity(
    cost: float,
    join_count: int,
    subquery_count: int,
) -> QueryComplexity:
    """Calculate query complexity based on metrics."""
    if cost <= COMPLEXITY_THRESHOLDS["simple"]["cost"] and join_count == 0 and subquery_count == 0:
        return QueryComplexity.SIMPLE
    elif cost <= COMPLEXITY_THRESHOLDS["moderate"]["cost"] and join_count <= 2 and subquery_count <= 1:
        return QueryComplexity.MODERATE
    elif cost <= COMPLEXITY_THRESHOLDS["complex"]["cost"] and join_count <= 5 and subquery_count <= 3:
        return QueryComplexity.COMPLEX
    else:
        return QueryComplexity.VERY_COMPLEX


def get_slow_query_severity(execution_time_ms: float) -> str:
    """Get severity level for a slow query."""
    if execution_time_ms >= SLOW_QUERY_THRESHOLDS["critical"]:
        return "critical"
    elif execution_time_ms >= SLOW_QUERY_THRESHOLDS["very_slow"]:
        return "very_slow"
    elif execution_time_ms >= SLOW_QUERY_THRESHOLDS["slow"]:
        return "slow"
    elif execution_time_ms >= SLOW_QUERY_THRESHOLDS["warning"]:
        return "warning"
    return "normal"


def hash_query(sql: str) -> str:
    """Generate a hash for a SQL query."""
    import hashlib
    import re

    # Normalize whitespace
    normalized = " ".join(sql.split())
    # Remove comments
    normalized = re.sub(r'--.*$', '', normalized, flags=re.MULTILINE)
    normalized = re.sub(r'/\*.*?\*/', '', normalized, flags=re.DOTALL)
    # Lowercase
    normalized = normalized.lower().strip()

    return hashlib.sha256(normalized.encode()).hexdigest()[:16]


def estimate_row_width(columns: list[dict[str, Any]]) -> int:
    """Estimate average row width based on column types."""
    type_sizes = {
        "integer": 4,
        "bigint": 8,
        "smallint": 2,
        "real": 4,
        "double": 8,
        "numeric": 16,
        "boolean": 1,
        "date": 4,
        "timestamp": 8,
        "text": 100,  # Estimate
        "varchar": 50,  # Estimate
        "json": 500,  # Estimate
        "uuid": 16,
    }

    total = 24  # Header overhead
    for col in columns:
        col_type = col.get("type", "text").lower()
        for type_name, size in type_sizes.items():
            if type_name in col_type:
                total += size
                break
        else:
            total += 50  # Default estimate

    return total

File: frontend/src/types/query_optimization.ts
/**
 * Query Optimization Types
 *
 * TypeScript types for query plans, cost estimation, optimization suggestions,
 * and performance monitoring.
 */

// Enums

export enum QueryPlanNodeType {
  SEQ_SCAN = 'seq_scan',
  INDEX_SCAN = 'index_scan',
  INDEX_ONLY_SCAN = 'index_only_scan',
  BITMAP_SCAN = 'bitmap_scan',
  NESTED_LOOP = 'nested_loop',
  HASH_JOIN = 'hash_join',
  MERGE_JOIN = 'merge_join',
  SORT = 'sort',
  AGGREGATE = 'aggregate',
  GROUP_BY = 'group_by',
  LIMIT = 'limit',
  SUBQUERY = 'subquery',
  CTE = 'cte',
  UNION = 'union',
  MATERIALIZE = 'materialize',
  FILTER = 'filter',
}

export enum QueryComplexity {
  SIMPLE = 'simple',
  MODERATE = 'moderate',
  COMPLEX = 'complex',
  VERY_COMPLEX = 'very_complex',
}

export enum OptimizationStatus {
  NOT_ANALYZED = 'not_analyzed',
  ANALYZING = 'analyzing',
  OPTIMIZED = 'optimized',
  NEEDS_ATTENTION = 'needs_attention',
  CRITICAL = 'critical',
}

export enum SuggestionPriority {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical',
}

export enum SuggestionCategory {
  INDEX = 'index',
  JOIN = 'join',
  FILTER = 'filter',
  AGGREGATION = 'aggregation',
  SUBQUERY = 'subquery',
  LIMIT = 'limit',
  CACHING = 'caching',
  SCHEMA = 'schema',
  STATISTICS = 'statistics',
}

export enum QuerySourceType {
  CHART = 'chart',
  DASHBOARD = 'dashboard',
  SQL_LAB = 'sql_lab',
  API = 'api',
  TRANSFORM = 'transform',
  SCHEDULED = 'scheduled',
  KODEE = 'kodee',
}

// Query Plan Types

export interface QueryPlanNode {
  id: string;
  type: QueryPlanNodeType;
  label: string;
  table_name?: string | null;
  index_name?: string | null;
  condition?: string | null;
  estimated_rows: number;
  actual_rows?: number | null;
  estimated_cost: number;
  actual_cost?: number | null;
  width: number;
  loops: number;
  startup_cost: number;
  total_cost: number;
  execution_time_ms?: number | null;
  children: QueryPlanNode[];
  warnings: string[];
  is_bottleneck: boolean;
}

export interface QueryPlan {
  id: string;
  query_hash: string;
  sql: string;
  connection_id: string;
  root_node: QueryPlanNode;
  total_cost: number;
  estimated_rows: number;
  actual_rows?: number | null;
  execution_time_ms?: number | null;
  planning_time_ms?: number | null;
  complexity: QueryComplexity;
  analyzed_at: string;
  metadata: Record<string, unknown>;
}

export interface QueryPlanRequest {
  sql: string;
  connection_id: string;
  explain_analyze?: boolean;
  explain_buffers?: boolean;
  explain_verbose?: boolean;
}

export interface QueryPlanComparison {
  original_plan: QueryPlan;
  optimized_plan: QueryPlan;
  cost_reduction_percent: number;
  row_estimate_diff: number;
  execution_time_improvement_ms?: number | null;
  improvements: string[];
}

// Cost Estimation Types

export interface CostComponent {
  name: string;
  description: string;
  estimated_cost: number;
  percentage: number;
  io_cost: number;
  cpu_cost: number;
}

export interface CostEstimate {
  query_hash: string;
  total_cost: number;
  io_cost: number;
  cpu_cost: number;
  startup_cost: number;
  row_estimate: number;
  width_estimate: number;
  data_transfer_mb: number;
  components: CostComponent[];
  complexity: QueryComplexity;
  estimated_duration_ms: number;
  confidence: number;
}

export interface CostEstimateRequest {
  sql: string;
  connection_id: string;
  include_components?: boolean;
}

export interface ResourceUsageEstimate {
  memory_mb: number;
  temp_space_mb: number;
  io_operations: number;
  cpu_time_ms: number;
  network_transfer_mb: number;
}

// Query Suggestion Types

export interface QuerySuggestion {
  id: string;
  category: SuggestionCategory;
  priority: SuggestionPriority;
  title: string;
  description: string;
  impact: string;
  original_snippet?: string | null;
  suggested_snippet?: string | null;
  estimated_improvement_percent: number;
  implementation_effort: string;
  auto_applicable: boolean;
}

export interface IndexSuggestion {
  table_name: string;
  schema_name: string;
  columns: string[];
  index_type: string;
  unique: boolean;
  partial_condition?: string | null;
  estimated_size_mb: number;
  estimated_improvement_percent: number;
  create_statement: string;
  priority: SuggestionPriority;
}

export interface QueryRewrite {
  original_sql: string;
  rewritten_sql: string;
  rewrite_type: string;
  description: string;
  estimated_improvement_percent: number;
  confidence: number;
}

export interface OptimizationResult {
  query_hash: string;
  original_sql: string;
  status: OptimizationStatus;
  complexity: QueryComplexity;
  cost_estimate: CostEstimate;
  suggestions: QuerySuggestion[];
  index_suggestions: IndexSuggestion[];
  rewrites: QueryRewrite[];
  bottlenecks: string[];
  analyzed_at: string;
}

// Query Execution Types

export interface QueryExecution {
  id: string;
  query_hash: string;
  sql: string;
  connection_id: string;
  source_type: QuerySourceType;
  source_id?: string | null;
  user_id?: string | null;
  workspace_id?: string | null;
  execution_time_ms: number;
  rows_returned: number;
  rows_scanned: number;
  bytes_processed: number;
  cached: boolean;
  cache_hit: boolean;
  status: string;
  error_message?: string | null;
  executed_at: string;
  metadata: Record<string, unknown>;
}

export interface QueryPerformanceStats {
  query_hash: string;
  sql_preview: string;
  execution_count: number;
  total_execution_time_ms: number;
  avg_execution_time_ms: number;
  min_execution_time_ms: number;
  max_execution_time_ms: number;
  p50_execution_time_ms: number;
  p95_execution_time_ms: number;
  p99_execution_time_ms: number;
  avg_rows_returned: number;
  total_rows_scanned: number;
  cache_hit_rate: number;
  error_rate: number;
  last_executed_at: string;
  first_executed_at: string;
}

export interface SlowQuery {
  id: string;
  query_hash: string;
  sql: string;
  connection_id: string;
  execution_time_ms: number;
  threshold_ms: number;
  rows_scanned: number;
  rows_returned: number;
  source_type: QuerySourceType;
  source_id?: string | null;
  user_id?: string | null;
  optimization_status: OptimizationStatus;
  suggestions_count: number;
  executed_at: string;
}

// Query History Types

export interface QueryHistoryEntry {
  id: string;
  query_hash: string;
  sql: string;
  connection_id: string;
  source_type: QuerySourceType;
  source_id?: string | null;
  user_id?: string | null;
  execution_time_ms: number;
  rows_returned: number;
  status: string;
  executed_at: string;
  tags: string[];
}

export interface QueryHistoryStats {
  total_queries: number;
  unique_queries: number;
  total_execution_time_ms: number;
  avg_execution_time_ms: number;
  slow_query_count: number;
  error_count: number;
  cache_hit_rate: number;
  by_source: Record<string, number>;
  by_hour: Array<Record<string, unknown>>;
  top_slow_queries: SlowQuery[];
}

// Configuration Types

export interface OptimizationConfig {
  slow_query_threshold_ms: number;
  auto_analyze: boolean;
  auto_suggest_indexes: boolean;
  max_suggestions_per_query: number;
  cache_plans: boolean;
  plan_cache_ttl_seconds: number;
  collect_statistics: boolean;
  statistics_sample_rate: number;
}

export interface OptimizationConfigUpdate {
  slow_query_threshold_ms?: number;
  auto_analyze?: boolean;
  auto_suggest_indexes?: boolean;
  max_suggestions_per_query?: number;
  cache_plans?: boolean;
  plan_cache_ttl_seconds?: number;
  collect_statistics?: boolean;
  statistics_sample_rate?: number;
}

// Response Types

export interface QueryPlanListResponse {
  plans: QueryPlan[];
  total: number;
}

export interface QuerySuggestionListResponse {
  suggestions: QuerySuggestion[];
  total: number;
}

export interface SlowQueryListResponse {
  queries: SlowQuery[];
  total: number;
}

export interface QueryHistoryListResponse {
  entries: QueryHistoryEntry[];
  total: number;
}

export interface IndexSuggestionListResponse {
  suggestions: IndexSuggestion[];
  total: number;
}

// Dashboard Types

export interface OptimizationDashboardSummary {
  total_queries_analyzed: number;
  slow_queries_count: number;
  critical_queries_count: number;
  avg_execution_time_ms: number;
  cache_hit_rate: number;
  pending_suggestions: number;
  recent_improvements: number;
  top_bottlenecks: string[];
}

export interface PerformanceTrend {
  date: string;
  total_queries: number;
  avg_execution_time_ms: number;
  slow_queries: number;
  cache_hit_rate: number;
}

// Constants

export const COMPLEXITY_LABELS: Record<QueryComplexity, string> = {
  [QueryComplexity.SIMPLE]: 'Simple',
  [QueryComplexity.MODERATE]: 'Moderate',
  [QueryComplexity.COMPLEX]: 'Complex',
  [QueryComplexity.VERY_COMPLEX]: 'Very Complex',
};

export const COMPLEXITY_COLORS: Record<QueryComplexity, string> = {
  [QueryComplexity.SIMPLE]: 'green',
  [QueryComplexity.MODERATE]: 'blue',
  [QueryComplexity.COMPLEX]: 'yellow',
  [QueryComplexity.VERY_COMPLEX]: 'red',
};

export const OPTIMIZATION_STATUS_LABELS: Record<OptimizationStatus, string> = {
  [OptimizationStatus.NOT_ANALYZED]: 'Not Analyzed',
  [OptimizationStatus.ANALYZING]: 'Analyzing',
  [OptimizationStatus.OPTIMIZED]: 'Optimized',
  [OptimizationStatus.NEEDS_ATTENTION]: 'Needs Attention',
  [OptimizationStatus.CRITICAL]: 'Critical',
};

export const OPTIMIZATION_STATUS_COLORS: Record<OptimizationStatus, string> = {
  [OptimizationStatus.NOT_ANALYZED]: 'gray',
  [OptimizationStatus.ANALYZING]: 'blue',
  [OptimizationStatus.OPTIMIZED]: 'green',
  [OptimizationStatus.NEEDS_ATTENTION]: 'yellow',
  [OptimizationStatus.CRITICAL]: 'red',
};

export const SUGGESTION_PRIORITY_LABELS: Record<SuggestionPriority, string> = {
  [SuggestionPriority.LOW]: 'Low',
  [SuggestionPriority.MEDIUM]: 'Medium',
  [SuggestionPriority.HIGH]: 'High',
  [SuggestionPriority.CRITICAL]: 'Critical',
};

export const SUGGESTION_PRIORITY_COLORS: Record<SuggestionPriority, string> = {
  [SuggestionPriority.LOW]: 'gray',
  [SuggestionPriority.MEDIUM]: 'blue',
  [SuggestionPriority.HIGH]: 'yellow',
  [SuggestionPriority.CRITICAL]: 'red',
};

export const SUGGESTION_CATEGORY_LABELS: Record<SuggestionCategory, string> = {
  [SuggestionCategory.INDEX]: 'Index',
  [SuggestionCategory.JOIN]: 'Join',
  [SuggestionCategory.FILTER]: 'Filter',
  [SuggestionCategory.AGGREGATION]: 'Aggregation',
  [SuggestionCategory.SUBQUERY]: 'Subquery',
  [SuggestionCategory.LIMIT]: 'Limit',
  [SuggestionCategory.CACHING]: 'Caching',
  [SuggestionCategory.SCHEMA]: 'Schema',
  [SuggestionCategory.STATISTICS]: 'Statistics',
};

export const QUERY_SOURCE_LABELS: Record<QuerySourceType, string> = {
  [QuerySourceType.CHART]: 'Chart',
  [QuerySourceType.DASHBOARD]: 'Dashboard',
  [QuerySourceType.SQL_LAB]: 'SQL Lab',
  [QuerySourceType.API]: 'API',
  [QuerySourceType.TRANSFORM]: 'Transform',
  [QuerySourceType.SCHEDULED]: 'Scheduled',
  [QuerySourceType.KODEE]: 'Kodee',
};

export const NODE_TYPE_LABELS: Record<QueryPlanNodeType, string> = {
  [QueryPlanNodeType.SEQ_SCAN]: 'Sequential Scan',
  [QueryPlanNodeType.INDEX_SCAN]: 'Index Scan',
  [QueryPlanNodeType.INDEX_ONLY_SCAN]: 'Index Only Scan',
  [QueryPlanNodeType.BITMAP_SCAN]: 'Bitmap Scan',
  [QueryPlanNodeType.NESTED_LOOP]: 'Nested Loop',
  [QueryPlanNodeType.HASH_JOIN]: 'Hash Join',
  [QueryPlanNodeType.MERGE_JOIN]: 'Merge Join',
  [QueryPlanNodeType.SORT]: 'Sort',
  [QueryPlanNodeType.AGGREGATE]: 'Aggregate',
  [QueryPlanNodeType.GROUP_BY]: 'Group By',
  [QueryPlanNodeType.LIMIT]: 'Limit',
  [QueryPlanNodeType.SUBQUERY]: 'Subquery',
  [QueryPlanNodeType.CTE]: 'CTE',
  [QueryPlanNodeType.UNION]: 'Union',
  [QueryPlanNodeType.MATERIALIZE]: 'Materialize',
  [QueryPlanNodeType.FILTER]: 'Filter',
};

export const SLOW_QUERY_THRESHOLDS = {
  warning: 500,
  slow: 1000,
  very_slow: 5000,
  critical: 30000,
};

// Helper Functions

export function formatExecutionTime(ms: number): string {
  if (ms < 1) return `${(ms * 1000).toFixed(0)}μs`;
  if (ms < 1000) return `${ms.toFixed(0)}ms`;
  if (ms < 60000) return `${(ms / 1000).toFixed(1)}s`;
  return `${(ms / 60000).toFixed(1)}m`;
}

export function formatCost(cost: number): string {
  if (cost < 1000) return cost.toFixed(2);
  if (cost < 1000000) return `${(cost / 1000).toFixed(1)}K`;
  return `${(cost / 1000000).toFixed(1)}M`;
}

export function formatRows(rows: number): string {
  if (rows < 1000) return rows.toString();
  if (rows < 1000000) return `${(rows / 1000).toFixed(1)}K`;
  return `${(rows / 1000000).toFixed(1)}M`;
}

export function getSlowQuerySeverity(executionTimeMs: number): string {
  if (executionTimeMs >= SLOW_QUERY_THRESHOLDS.critical) return 'critical';
  if (executionTimeMs >= SLOW_QUERY_THRESHOLDS.very_slow) return 'very_slow';
  if (executionTimeMs >= SLOW_QUERY_THRESHOLDS.slow) return 'slow';
  if (executionTimeMs >= SLOW_QUERY_THRESHOLDS.warning) return 'warning';
  return 'normal';
}

export function getSlowQuerySeverityColor(severity: string): string {
  switch (severity) {
    case 'critical':
      return 'red';
    case 'very_slow':
      return 'orange';
    case 'slow':
      return 'yellow';
    case 'warning':
      return 'blue';
    default:
      return 'green';
  }
}

export function getComplexityColor(complexity: QueryComplexity): string {
  return COMPLEXITY_COLORS[complexity] || 'gray';
}

export function getOptimizationStatusColor(status: OptimizationStatus): string {
  return OPTIMIZATION_STATUS_COLORS[status] || 'gray';
}

export function getPriorityColor(priority: SuggestionPriority): string {
  return SUGGESTION_PRIORITY_COLORS[priority] || 'gray';
}

export function calculateImprovement(original: number, optimized: number): number {
  if (original === 0) return 0;
  return ((original - optimized) / original) * 100;
}

export function isBottleneckNode(node: QueryPlanNode): boolean {
  return (
    node.is_bottleneck ||
    node.type === QueryPlanNodeType.SEQ_SCAN ||
    (node.estimated_rows > 10000 && node.type === QueryPlanNodeType.NESTED_LOOP)
  );
}

export function getNodeWarnings(node: QueryPlanNode): string[] {
  const warnings: string[] = [...node.warnings];

  if (node.type === QueryPlanNodeType.SEQ_SCAN && node.estimated_rows > 1000) {
    warnings.push('Sequential scan on large table - consider adding an index');
  }

  if (node.actual_rows != null && node.estimated_rows > 0) {
    const ratio = node.actual_rows / node.estimated_rows;
    if (ratio > 10 || ratio < 0.1) {
      warnings.push('Row estimate significantly different from actual - statistics may be stale');
    }
  }

  return warnings;
}

// State Management

export interface QueryOptimizationState {
  plans: QueryPlan[];
  currentPlan: QueryPlan | null;
  suggestions: QuerySuggestion[];
  indexSuggestions: IndexSuggestion[];
  slowQueries: SlowQuery[];
  history: QueryHistoryEntry[];
  historyStats: QueryHistoryStats | null;
  config: OptimizationConfig | null;
  dashboardSummary: OptimizationDashboardSummary | null;
  performanceTrends: PerformanceTrend[];
  isLoading: boolean;
  isAnalyzing: boolean;
  error: string | null;
}

export function createInitialQueryOptimizationState(): QueryOptimizationState {
  return {
    plans: [],
    currentPlan: null,
    suggestions: [],
    indexSuggestions: [],
    slowQueries: [],
    history: [],
    historyStats: null,
    config: null,
    dashboardSummary: null,
    performanceTrends: [],
    isLoading: false,
    isAnalyzing: false,
    error: null,
  };
}
